gnupg-pkcs11-scd/command.c

```diff
@@ -1134,127 +1201,169 @@ gpg_error_t _cmd_pksign_type (assuan_context_t ctx, char *line, int typehint)
+				inject = INJECT_NONE;
+			}
+			else if (!strcmp(hash, "sha512") && data->size == 0x40) {
+				inject = INJECT_SHA512;
+			}
+			else {
+				common_log (LOG_DEBUG, "unsupported hash algo (hash=%s,size=%d)", hash, data->size);
 				error = GPG_ERR_UNSUPPORTED_ALGORITHM;
 				goto cleanup;
 			}
 		}
 		else {
-			/*
-			 * unknown hash algorithm;
-			 * gnupg's scdaemon forces to SHA1
-			 */
-			inject = INJECT_SHA1;
-
-			/* When doing auth operation, hash algorithm prefix detection does not work
-			 * but data always comes with algorithm appended, so do not append anything
-			 * by default. */
-			if (typehint == OPENPGP_AUTH) {
-				inject = INJECT_NONE;
+			if (
+				data->size == 0x10 + sizeof (md5_prefix_pkcs1) ||
+				data->size == 0x14 + sizeof (sha1_prefix_pkcs1) ||
+				data->size == 0x14 + sizeof (rmd160_prefix_pkcs1)
+			) {
+				if (
+					memcmp (data->data, md5_prefix_pkcs1, sizeof (md5_prefix_pkcs1)) &&
+					memcmp (data->data, sha1_prefix_pkcs1, sizeof (sha1_prefix_pkcs1)) &&
+					memcmp (data->data, rmd160_prefix_pkcs1, sizeof (rmd160_prefix_pkcs1))
+				) {
+					error = GPG_ERR_UNSUPPORTED_ALGORITHM;
+					goto cleanup;
+				}
+			}
+			else {
+				/*
+				 * unknown hash algorithm;
+				 * gnupg's scdaemon forces to SHA1
+				 */
+				inject = INJECT_SHA1;
+
+				/* When doing auth operation, hash algorithm prefix detection does not work
+				 * but data always comes with algorithm appended, so do not append anything
+				 * by default. */
+				if (typehint == OPENPGP_AUTH) {
+					inject = INJECT_NONE;
+				}
 			}
 		}
+	} else {
+		/* Non-PKCS1 does not inject anything */
+		inject = INJECT_NONE;
 	}
 
 	if (inject != INJECT_NONE) {
 		const unsigned char *oid;
 		size_t oid_size;
 		switch (inject) {
 			case INJECT_RMD160:
-				oid = rmd160_prefix;
-				oid_size = sizeof (rmd160_prefix);
+				oid = rmd160_prefix_pkcs1;
+				oid_size = sizeof (rmd160_prefix_pkcs1);
 			break;
 			case INJECT_MD5:
-				oid = md5_prefix;
-				oid_size = sizeof (md5_prefix);
+				oid = md5_prefix_pkcs1;
+				oid_size = sizeof (md5_prefix_pkcs1);
 			break;
 			case INJECT_SHA1:
-				oid = sha1_prefix;
-				oid_size = sizeof (sha1_prefix);
+				oid = sha1_prefix_pkcs1;
+				oid_size = sizeof (sha1_prefix_pkcs1);
 			break;
 			case INJECT_SHA224:
-				oid = sha224_prefix;
-				oid_size = sizeof (sha224_prefix);
+				oid = sha224_prefix_pkcs1;
+				oid_size = sizeof (sha224_prefix_pkcs1);
 			break;
 			case INJECT_SHA256:
-				oid = sha256_prefix;
-				oid_size = sizeof(sha256_prefix);
+				oid = sha256_prefix_pkcs1;
+				oid_size = sizeof(sha256_prefix_pkcs1);
 			break;
 			case INJECT_SHA384:
-				oid = sha384_prefix;
-				oid_size = sizeof(sha384_prefix);
+				oid = sha384_prefix_pkcs1;
+				oid_size = sizeof(sha384_prefix_pkcs1);
 			break;
 			case INJECT_SHA512:
-				oid = sha512_prefix;
-				oid_size = sizeof(sha512_prefix);
+				oid = sha512_prefix_pkcs1;
+				oid_size = sizeof(sha512_prefix_pkcs1);
```

Comment:
renames should be done in own pull request.

Reply:
This is no longer renamed since the PKCS1 prefixes have been moved entirely.